lib/identity_cache/cached/prefetcher.rb

```diff
@@ -44,155 +44,8 @@ def fetch_association(klass, association, records)
-                parent_id = child_record.send(reflection.foreign_key)
-                if parent_id && !child_record.instance_variable_defined?(cached_iv_name)
-                  hash[parent_id] = child_record
-                end
-              end
-              parent_records = reflection.klass.fetch_multi(ids_to_child_record.keys)
-              parent_records.each do |parent_record|
-                child_record = ids_to_child_record[parent_record.id]
-                child_record.instance_variable_set(cached_belongs_to.records_variable_name, parent_record)
-              end
-            end
-          end
-
-          parent_records
-        end
-
-        def prefetch_embedded(cached_association, records)
-          # Make the same assumption as ActiveRecord::Associations::Preloader, which is
-          # that all the records have the same associations loaded, so we can just check
-          # the first record to see if an association is loaded.
-          first_record = records.first
-          return if first_record.association(cached_association.name).loaded?
-          if cached_association.embedded_recursively?
-            return if first_record.instance_variable_defined?(cached_association.dehydrated_variable_name)
-            return if first_record.instance_variable_defined?(cached_association.records_variable_name)
-          elsif first_record.instance_variable_defined?(cached_association.ids_variable_name)
-            return
-          end
-          fetch_embedded(cached_association, records)
-        end
-
-        def fetch_embedded(cached_association, records)
-          klass = cached_association.reflection.active_record
-          associations = klass.send(:embedded_associations)
-          return if associations.empty?
-
-          return unless klass.primary_cache_index_enabled
-
-          cached_records_by_id = klass.fetch_multi(records.map(&:id)).index_by(&:id)
-
-          associations.each_value do |association|
-            records.each do |record|
-              next unless (cached_record = cached_records_by_id[record.id])
-              case association
-              when Cached::Reference::HasMany
-                cached_association = cached_record.public_send(association.cached_ids_name)
-                record.instance_variable_set(association.ids_variable_name, cached_association)
-              when Cached::Reference::HasOne
-                cached_association = cached_record.public_send(association.cached_id_name)
-                record.instance_variable_set(association.id_variable_name, cached_association)
-              else
-                cached_association = cached_record.public_send(association.cached_accessor_name)
-                record.instance_variable_set(association.records_variable_name, cached_association)
-              end
-            end
-          end
+          cached_association = klass.send(:cached_association, association)
```

Comment:
Why are we using `send` for a method that we own? If it is to "tell" people this method is private API, documentation would be better. We can say that people will ignore the documentation, but if they want to use this method even if the documentation say it is private they will likely also use `send`.

Reply:
I'd be in favour of making this a nodoc public method, but I was planning on doing this later. I'm trying my best to improve IDC's codebase little by little 🙃

Reply:
https://github.com/Shopify/identity_cache/pull/410